File: front_end/front_end/mock/data/contest.ts
import { random } from 'lodash-es'
import type { ACMCell, Contest, Ranking, Signup } from '../interface'
import { mockUsers } from './user'

// 比赛测试数据
export const mockContests: Contest[] = [
  {
    id: 1,
    title: '2023年春季编程大赛我是一个名字很长很长很长很长很长很长很长很长很长很长很长很长很的比赛',
    banner: 'https://example.com/contest1-banner.jpg',
    summary: '年度大型编程竞赛，涵盖算法与数据结构',
    description: `# 2023年春季编程大赛\n\n## 比赛规则\n\n- 采用ICPC赛制\n- 5道题目，难度递增\n- 排名按解题数，同解题数按罚时\n\n## 奖励\n\n- 第一名：奖金5000元\n- 第二名：奖金3000元\n- 第三名：奖金1000元`,
    difficulty: 4,
    status: 'public',
    startTime: new Date(Date.now() - 2 * 60 * 60 * 1000), // 已开始 2 小时
    endTime: new Date(Date.now() + 3 * 60 * 60 * 1000), // 持续 5 小时
    rule: 'OI',
    problems: {
      A: 1001,
      B: 1002,
      C: 1003,
      D1: 1004,
      D2: 1005,
      E: 1006,
      F: 1007,
      G: 1008,
      H: 1009,
      I: 1010,
    },
    createdAt: new Date('2023-01-15'),
    updatedAt: new Date('2023-01-15'),
  },
  {
    id: 2,
    title: '算法入门挑战赛',
    banner: 'https://example.com/contest2-banner.jpg',
    summary: '适合初学者的算法竞赛',
    description: `# 算法入门挑战赛\n\n适合刚接触算法竞赛的选手参加`,
    difficulty: 2,
    status: 'public',
    startTime: new Date(Date.now() - 2 * 24 * 60 * 60 * 1000), // 2天前开始
    endTime: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000), // 已结束
    rule: 'ACM',
    problems: {
      A: 1001,
      B: 1004,
    },
    createdAt: new Date('2023-02-10'),
    updatedAt: new Date('2023-02-10'),
  },
  {
    id: 3,
    title: 'ACM校队选拔赛',
    banner: 'https://example.com/contest3-banner.jpg',
    summary: '校ACM队选拔专用比赛',
    description: '仅限校内学生参加',
    difficulty: 5,
    status: 'private',
    startTime: new Date(Date.now() + 5 * 24 * 60 * 60 * 1000), // 5 天后开始
    endTime: new Date(Date.now() + 6 * 24 * 60 * 60 * 1000), // 持续 1 天
    rule: 'IOI',
    problems: {
      A: 1001,
      B1: 1002,
      B2: 1004,
    },
    createdAt: new Date('2023-03-01'),
    updatedAt: new Date('2023-03-01'),
  },
]

// 比赛报名数据
export const mockRegistrations: Signup[] = [
  {
    contest: 1,
    user: 1,
    stamp: new Date('2023-02-09'),
  },
  {
    contest: 2,
    user: 2,
    stamp: new Date('2023-02-09'),
  },
]

// 比赛排名数据
export const mockRankings: Ranking[] = [
  {
    ranking: 1,
    contest: 2,
    user: 1,
    detail: {
      type: 'ACM',
      totalPenalty: 500,
      totalSolved: 5,
      problems: {
        1001: generateACMCell(),
        1002: generateACMCell(),
        1003: generateACMCell(),
        1004: generateACMCell(),
      }
    }
  },
  {
    ranking: 1,
    contest: 2,
    user: 2,
    detail: {
      type: 'ACM',
      totalPenalty: 500,
      totalSolved: 5,
      problems: {
        1001: generateACMCell(),
        1002: generateACMCell(),
        1003: generateACMCell(),
        1004: generateACMCell(),
      }
    }
  },
]

export function generateACMCell() {
  const cell: ACMCell = {
    isFirst: Math.random() < 0.1,   // 是否为一血
    isSolved: Math.random() < 0.5,  // 是否已通过
    attemptBF: random(0, 10),  // 封榜前尝试次数
    attemptAF: random(0, 10),  // 封榜后尝试次数
    penalty: random(1, 100),  // 罚时
  }
  return cell
}

export function generateRanklist() {
  const ranklist: Ranking[] = []
  for (let i = 0; i < 200; ++i) {
    const item: Ranking = {
      contest: 1,
      user: random(1, mockUsers.length),
      ranking: i + 1,
      detail: {
        type: 'ACM',
        totalPenalty: 500,
        totalSolved: 5,
        problems: {}
      }
    }
    for (let j = random(1, 20); j >= 0; --j) {
      item.detail.problems[random(1000, 1050)] = generateACMCell()
    }
    ranklist.push(item)
  }
  return ranklist
}

File: back_end/back_end/internal/model/contest.go
package model

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"

	"gorm.io/datatypes"
)

type ContestDifficulty int
type ContestStatus string
type ContestRule string

const (
	ContestDifficulty1 ContestDifficulty = 1
	ContestDifficulty2 ContestDifficulty = 2
	ContestDifficulty3 ContestDifficulty = 3
	ContestDifficulty4 ContestDifficulty = 4
	ContestDifficulty5 ContestDifficulty = 5

	ContestStatusPrivate ContestStatus = "private"
	ContestStatusPublic  ContestStatus = "public"
	ContestStatusDeleted ContestStatus = "deleted"

	ContestRuleOI  ContestRule = "OI"
	ContestRuleACM ContestRule = "ACM"
	ContestRuleIOI ContestRule = "IOI"
)

type ContestProblems map[ProblemLabel]ProblemId

func (c *ContestProblems) Scan(value interface{}) error {
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}
	return json.Unmarshal(bytes, c)
}

func (c ContestProblems) Value() (driver.Value, error) {
	return json.Marshal(c)
}

type ContestProblemStatus struct {
	FirstBloodUserID *UserId    `json:"firstBloodUserId,omitempty"`
	FirstBloodTime   *time.Time `json:"firstBloodTime,omitempty"`
	SolvedCount      int        `json:"solvedCount"`
	TotalCount       int        `json:"totalCount"`
}

type ContestProblemStatuses map[ProblemId]ContestProblemStatus
func (c *ContestProblemStatuses) Scan(value interface{}) error {
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}
	return json.Unmarshal(bytes, c)
}

func (c ContestProblemStatuses) Value() (driver.Value, error) {
	return json.Marshal(c)
}

// 比赛基本信息
type Contest struct {
	BaseModel
	ID            ContestId         `json:"id"`
	Title         string            `gorm:"size:100"         json:"title"`
	Banner        string            `gorm:"size:200"         json:"banner,omitempty"`
	Summary       string            `gorm:"type:text"        json:"summary"`
	Description   string            `gorm:"type:text"        json:"description"`
	Difficulty    ContestDifficulty `gorm:"not null"         json:"difficulty"`
	Status        ContestStatus     `gorm:"type:varchar(10)" json:"status"`
	StartTime     time.Time         `gorm:"not null"         json:"startTime"`
	EndTime       time.Time         `gorm:"not null"         json:"endTime"`
	Rule          ContestRule       `gorm:"type:varchar(10)" json:"rule"`
	Problems      ContestProblems   `gorm:"type:json"        json:"problems"`
	ProblemStatus ContestProblemStatuses `gorm:"type:json" json:"problemStatus,omitempty"`
}

// 比赛报名信息
type Signup struct {
	ContestID ContestId `gorm:"primaryKey"     json:"contest"`
	UserID    UserId    `gorm:"primaryKey"     json:"user"`
	Stamp     time.Time `gorm:"autoCreateTime" json:"register"`
}

func (Signup) TableName() string {
	return "signups"
}

// 用于比赛列表查询
type ContestWithSignups struct {
	Contest
	Signups []Signup `gorm:"foreignKey:ContestID" json:"signups"` // 反向关联
}

func (ContestWithSignups) TableName() string {
	return "contests"
}

// 比赛排名信息
type Ranking struct {
	ContestID ContestId      `gorm:"primaryKey"  json:"contest"`
	UserID    UserId         `gorm:"primaryKey"  json:"user"`
	Team      string         `gorm:"size:50"     json:"team"`
	Ranking   int            `                   json:"ranking"`
	Detail    datatypes.JSON `gorm:"type:json"   json:"detail"`
}

func (Ranking) TableName() string {
	return "rankings"
}

// ACM problem cell data
type ACMCell struct {
	IsFirst   bool `json:"isFirst"`   // 是否为一血
	IsSolved  bool `json:"isSolved"`  // 是否已通过
	AttemptBF int  `json:"attemptBF"` // 封榜前尝试次数
	AttemptAF int  `json:"attemptAF"` // 封榜后尝试次数
	Penalty   int  `json:"penalty"`   // 罚时
}

// ACM ranking detail
type ACMDetail struct {
	Type         string                `json:"type"`         // "ACM"
	TotalPenalty int                   `json:"totalPenalty"` // 总罚时
	TotalSolved  int                   `json:"totalSolved"`  // 总通过数
	Problems     map[ProblemId]ACMCell `json:"problems"`
}

// OI problem cell data
type OIProblem struct {
	Score int `json:"score"`
}

// OI ranking detail
type OIDetail struct {
	Type       string                  `json:"type"`       // "OI"
	TotalScore int                     `json:"totalScore"` // 总分
	Problems   map[ProblemId]OIProblem `json:"problems"`
}

// IOI problem cell data (same as OI for now)
type IOIProblem struct {
	Score int `json:"score"`
}

// IOI ranking detail
type IOIDetail struct {
	Type       string                   `json:"type"`       // "IOI"
	TotalScore int                      `json:"totalScore"` // 总分
	Problems   map[ProblemId]IOIProblem `json:"problems"`
}

// 比赛过滤条件
type ContestFilter struct {
	Status     *ContestStatus
	Rule       *ContestRule
	Difficulty *ContestDifficulty
	UserID     *UserId // 用户是否已报名
	Keyword    *string
	Before     *time.Time // 开始时间之前
	After      *time.Time // 开始时间之后
}

// 原始过滤参数
type ContestFilterRaw struct {
	Status     *ContestStatus     `json:"status,omitempty"`
	Rule       *ContestRule       `json:"rule,omitempty"`
	Difficulty *ContestDifficulty `json:"difficulty,omitempty"`
	User       *string            `json:"user,omitempty"`
	Keyword    *string            `json:"keyword,omitempty"`
	Before     *time.Time         `json:"before,omitempty"`
	After      *time.Time         `json:"after,omitempty"`
}

// 比赛详情请求
type ContestRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛详情响应
type ContestResponse struct {
	Contest Contest  `json:"contest"`
	Signup  *Signup  `json:"signup,omitempty"`
	Ranking *Ranking `json:"ranking,omitempty"`
}

// 比赛试题请求
type ContestProblemsRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛试题响应
type ContestProblemsResponse struct {
	Problems []ProblemCore `json:"problems"`
}

// 比赛列表请求
type ContestListRequest struct {
	ContestFilterRaw
	Page *int `json:"page,omitempty"`
}

// 比赛列表响应
type ContestListResponse struct {
	Total    int64                `json:"total"`
	Contests []ContestWithSignups `json:"contests"`
}

// 比赛列表请求
type ContestAllRequest struct {
	ContestFilterRaw
	Page *int `json:"page,omitempty"`
	Size *int `json:"size,omitempty"`
}

// 比赛列表响应
type ContestAllResponse struct {
	Total    int64     `json:"total"`
	Contests []Contest `json:"contests"`
}

// 比赛编辑请求
type ContestEditRequest struct {
	Contest Contest `json:"contest"`
}

// 比赛编辑响应
type ContestEditResponse struct {
	Contest Contest `json:"contest"`
}

// 比赛删除请求
type ContestDeleteRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛删除响应
type ContestDeleteResponse struct {
}

// 比赛报名请求
type ContestSignupRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛报名响应
type ContestSignupResponse struct {
}

// 比赛取消报名请求
type ContestSignoutRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛取消报名响应
type ContestSignoutResponse struct {
}

// 比赛排名请求
type ContestRankingRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛排名响应
type ContestRankingResponse struct {
	Ranking *Ranking `json:"ranking,omitempty"`
}

// 比赛排行榜请求
type ContestRanklistRequest struct {
	Contest ContestId `json:"contest"`
}

// 比赛排行榜响应
type ContestRanklistResponse struct {
	Rankings []Ranking `json:"rankings"`
}
